### PR TITLE
feat(ng-form): add ng form library

### PR DESCRIPTION
**Pull request by ivan-morandi-zupit**
## Checklist before requesting a review
- [ ] I have performed a self-review of my code
- [ ] Does the code follow the team's coding standards and best practices?
- [ ] Is the code well-organized and easy to understand?
- [ ] Are file, variable and function names clear and meaningful?
- [ ] Are there any duplicated code blocks that can be refactored into reusable functions?

## Functionality
- [ ] Does the PR implement the intended functionality or fix the issue described in the title and description?
- [ ] Are there any edge cases that have been overlooked?

## Testing
- [ ] Are there sufficient tests covering the changes?
- [ ] Are there any missing test cases that need to be added?
- [ ] Has the code been tested in different environments if applicable (e.g., browsers, platforms, devices)?

## Documentation
- [ ] Are the changes well-documented and easy to understand?
- [ ] Are there any missing or outdated comments in the code?
- [ ] Is the PR title clear and detailed enough to provide necessary context?

## Performance
- [ ] Do the changes have any impact on the system's performance?
- [ ] Are there any potential bottlenecks that need to be addressed?

## Dependencies
- [ ] Have the changes introduced any new dependencies or updated existing ones?
- [ ] Are the dependencies up-to-date and secure?

## Compatibility
- [ ] Will the changes have any adverse effects on other parts of the system?
- [ ] Do they work well with other existing features?

## Continuous Integration/Deployment (CI/CD)
- [ ] Have the changes passed the CI/CD pipeline?
- [ ] Are there any deployment considerations or steps that need to be taken into account?

## Code Review language
- [ ] Is the feedback provided in a constructive and respectful manner?
- [ ] Has the reviewer avoided personal attacks or harsh language?
- [ ] Are there any suggestions for improvement or alternative approaches?

### CODE REVIEW

File: packages/ng-form/src/lib/utils/errors.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { ERROR_EXPECTED_MISSING_VALUE } from '../config'
+import { inject } from '@angular/core'
+
+export class FormException extends Error {
```
Comment (ivan-morandi-zupit) [inline packages/ng-form/src/lib/utils/errors.ts:4]: visto che in alcuni casi gestivamo i controll con getValueOrError ho deciso di aggiungerlo anche qua, specifico però per FormException.

Avevo in mente 2 soluzioni alternative: o creare un'altra lib per gestire il getValueOrError o gestire in maniera alternativa. Entrambe le 2 soluzioni non mi ispiravano: tempo nel fare un'altra libreria eccessivo per poche righe di codice nel primo caso e valore perso dei controlli nel secondo. Per questo ho preferito importare direttamente qua la gestione degli errori per i form. Se non vi convince fatemi sapere

File: packages/ng-form/src/lib/utils/form.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  if (form.valid) {
+    return FormValidationResult.VALID
+  }
+  if (hasAnyRequiredErrors(form)) {
+    return FormValidationResult.MISSING_REQUIRED
+  }
+  return FormValidationResult.INVALID
+}
+
+export function validateFormWithException(
```
Comment (ivan-morandi-zupit) [inline packages/ng-form/src/lib/utils/form.ts:74]: mancano alcuni validatori presenti nel progetto più recente, ma mi sembravano troppo specifici al progetto e/o complessi che vanno bene solo per alcuni casi specifici

File: packages/ng-form/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# ng-form
+
+An Angular library to ease the management of Angular Forms.
```
Comment (ivan-morandi-zupit) [inline packages/ng-form/README.md:3]: The doc is not completed yet, it misses the description of Validators and Utilities. For know we keep the doc as it is as I need to move to another project

File: packages/ng-form/src/lib/components/form/form.component.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<form
+  [formGroup]="form"
+  [class]="formStyleClass"
+  (ngFormSubmit)="submitFormSave()">
```
Comment (alessio-libardi-zupit) [inline packages/ng-form/src/lib/components/form/form.component.html:4]: Not sure this is the way 🤔 why don't we let the use do what he needs to do with the `<form>` tag? Is there a reason for us to bring it within this library? What I expected were directives to be used within our applications.

File: packages/ng-form/src/lib/components/form/form.component.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  <div [class]="contentStyleClass">
+    <ng-content select="[content]"></ng-content>
+  </div>
+  <div [class]="footerStyleClass" *ngIf="showFooter">
+    <ng-content select="[footer]"></ng-content>
+  </div>
```
Comment (alessio-libardi-zupit) [inline packages/ng-form/src/lib/components/form/form.component.html:10]: Not sure this logic should be here 🤔 I expected the `ng-form` library to just give the utilities to better work with form, not to implement the layout of my app.

File: packages/ng-form/src/lib/components/form/form.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing'
```
Comment (alessio-libardi-zupit) [inline packages/ng-form/src/lib/components/form/form.component.spec.ts:1]: In general, it would be cool to use [angular-testing-library](https://testing-library.com/docs/angular-testing-library/intro/). Maybe I'll rewrite them myself using that library, which IMO is far better than `TestBed`

File: packages/ng-form/src/lib/components/form/form.component.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  Input,
+  OnChanges,
+  Output,
+  SimpleChanges
+} from '@angular/core'
+import { FormGroup } from '@angular/forms'
+import { setFormEnabled } from '../../utils/form'
+
+@Component({
+  selector: 'ng-form-form',
```
Comment (alessio-libardi-zupit) [inline packages/ng-form/src/lib/components/form/form.component.ts:14]: I don't imagine being happy whilst using this component in my templates:

`<ng-form-form>` 👿

File: packages/ng-form/src/lib/components/form/form.component.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    const formValue = this.getValue()
+    this.formSubmit.emit(formValue)
+  }
+
+  getValue(): Model {
+    return this.formValueToModelMapper(this.form.getRawValue())
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  ngOnChanges(changes: SimpleChanges): void {
```
Comment (alessio-libardi-zupit) [inline packages/ng-form/src/lib/components/form/form.component.ts:71]: Using `_` will prevent the eslint error

**Review by alessio-libardi-zupit — CHANGES_REQUESTED** [review]

File: packages/ng-form/src/lib/components/form/form.component.spec.ts (hunk 6 above)
Comment (ivan-morandi-zupit) [reply packages/ng-form/src/lib/components/form/form.component.spec.ts:1]: I tried it... but it seemed that for this case it was complicating the tests. I spent 30 mins without getting anything good. Maybe it's my fault and I was doing the tests in the wrong way with the library...

Comment (alessio-libardi-zupit) [reply packages/ng-form/src/lib/components/form/form.component.spec.ts:1]: Maybe its just better `TestBed` 🤷🏻‍♂️ I assumed `testing-library` was better because of the experience I had with both Cypress and React, but maybe that's not the case with Angular. Thanks for trying it out!

**Review by alessio-libardi-zupit — APPROVED** [review]

**Review by alessandro-mariotti-zupit — APPROVED** [review]
